shared/markdown/parser.pegjs

```diff
@@ -0,0 +1,111 @@
+start
+ = children:(Blank / Code / Content / Blank)* { return {type: 'text', children: children}; }
+
+Code = CodeBlock / InlineCode
+
+Content =
+	StyledText / Text
+
+StyledText
+ = QuoteBlock / Italic / Bold / Strike / Emoji
+
+// Define what our markers look like
+Ticks1 = "`" ! '`'
+Ticks3 = __? "```" __? ! '```'
+
+StrikeMarker = "~" ! "~"
+BoldMarker = "*" ! "*"
+ItalicMarker = "_" ! "_"
+EmojiMarker = ":" ! ":"
+QuoteBlockMarker = ">"
+
+// Define what we can go to when we are inside a style. e.g. Bold -> Bold doesn't make sense, but Bold -> Strike does
+FromBold
+ = (Italic / Strike / InsideBoldMarker)
+
+FromItalic
+ = Bold / Strike / InsideItalicMarker
+
+FromStrike
+ = Italic / Bold / InsideStrikeMarker
+
+FromQuote
+ = Italic / Bold / Strike / InsideQuoteBlock
+
+// Define what text inside a style looks like. Usually everything but the end marker
+InsideBoldMarker
+ = (! BoldMarker .) { return text(); }
+
+InsideItalicMarker
+ = ((! ItalicMarker) .) { return text(); }
+
+InsideStrikeMarker
+ = ((! StrikeMarker) .) { return text(); }
+
+InsideCodeBlock
+ = ((! Ticks3) .) { return text(); }
+
+InsideInlineCode
+ = ((! Ticks1) .) { return text(); }
+
+InsideQuoteBlock
+ = ((! LineTerminatorSequence) .) { return text(); }
+
+// Here we use the literal ":" because we want to not match the :foo in ::foo
+InsideEmojiMarker
+ = (! ":" .) { return text(); }
+
+// Define the rules for styles. Usually a start marker, children, and an end marker.
+QuoteBlock
+ = QuoteBlockMarker _? children:FromQuote* LineTerminatorSequence { return {type: 'quote-block', children: children}; }
+
+Bold
+ = BoldMarker children:FromBold* BoldMarker { return {type: 'bold', children: children}; }
+
+Italic
+ = ItalicMarker children:FromItalic* ItalicMarker { return {type: 'italic', children: children}; }
+
+Strike
+ = StrikeMarker children:FromStrike* StrikeMarker { return {type: 'strike', children: children}; }
+
+CodeBlock
+ = Ticks3 children:InsideCodeBlock* Ticks3 { return {type: 'code-block', children}; }
+
+InlineCode
+ = Ticks1 children:InsideInlineCode* Ticks1 { return {type: 'inline-code', children}; }
+
+Emoji
+ = EmojiMarker children:(InsideEmojiMarker / "::")* EmojiMarker { return {type: 'emoji', children: [children.join('')]}; }
```

Comment:
Does/should this also include unicode emoji? We should parse them so we can use the `Emoji` component on them, right?